### PR TITLE
[client] Add generic client (v2)

### PR DESCRIPTION
**Pull request by valeriocos**
This PR proposes a generic client (and replaces PR #208)  to deal with common problems (server errors, rate limit bans, connection losts) when fetching data from APIs.
The generic client has been successfully applied to the GitHub backend.

### CODE REVIEW

File: perceval/client.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+    def _fetch(self, url, payload=None, headers=None, method=GET,
+               stream=False,
+               retry_on_status_codes=None,
+               retry_after_status_codes=None):
+
+        self.set_retry_after_status_codes(retry_after_status_codes)
+        self.sleep_for_rate_limit()
+
+        session = self.session or requests.Session()
```
Comment (sduenas) [inline perceval/client.py]: Why doesn't the client use always a session?

I would have seen this code in a different method, maybe as part of the initialization of the instance, to call it just once.

File: perceval/client.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+            logger.debug("Rate limit: %s", self.rate_limit)
+        else:
+            self.rate_limit = None
+            self.rate_limit_reset_ts = None
+
+    def __send_request(self, session, url, payload=None, headers=None, method=GET, stream=False):
+        if method == self.GET:
+            response = session.get(url, params=payload, headers=headers, stream=stream)
+        else:
+            response = self.session.post(url, data=payload, headers=headers, stream=stream)
```
Comment (sduenas) [inline perceval/client.py]: I think this should be either `session` or `self.session`, but the same on the two calls. Moreover, the code should be something like this:

```
if method == self.GET:
    freq = session.get
else:
    freq = session.post

response = freq(url,  data=payload, headers=headers, stream=stream)
```

File: perceval/client.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    def create_http_session(self, headers=None):
+        self.session = requests.Session()
+
+        if headers:
+            self.session.headers.update(headers)
+
+    def close_http_session(self, session):
+        session.close()
+
+    def set_retry_after_status_codes(self, status_codes=None):
```
Comment (sduenas) [inline perceval/client.py]: Why is this needed?

File: perceval/client.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    def __init__(self, base_url, sleep_for_rate=False, min_rate_to_sleep=MIN_RATE_LIMIT,
+                 max_retries=MAX_RETRIES, default_sleep_time=DEFAULT_SLEEP_TIME):
+        self.base_url = base_url
+        self.rate_limit = None
+        self.rate_limit_reset_ts = None
+        self.session = None
+
+        self.sleep_for_rate = sleep_for_rate
+
+        if min_rate_to_sleep > self.MAX_RATE_LIMIT:
```
Comment (sduenas) [inline perceval/client.py]: I'm still not sure what to do with rate limit handlers. Not sure if we should move this behaviour to a specific class and use it as a decorator or to have a mixin. For instance:
```
class GitHubClient(RateLimitHandler, HttpClient):
    [...]
```

Comment (valeriocos) [reply perceval/client.py]: We can discuss about it. In your example, the responsabilities of the the HttpClient would be just about connection issues?

File: perceval/client.py (hunk 3 above)
Comment (valeriocos) [reply perceval/client.py]: set_retry_after_status_codes is needed to modify the list of status codes were the retry-after header may appear. By default, those codes are 503, 413, 429, however in same cases (GitHub), the retry-after can appear also with other status codes (403)

File: perceval/client.py (hunk 1 above)
Comment (valeriocos) [reply perceval/client.py]: From the [doc](http://docs.python-requests.org/en/latest/user/advanced/#session-objects), I read that "Session object allows you to persist certain parameters across requests". I guessed that we didn't need it and it was fine to create a session per get request.

File: perceval/client.py (hunk 2 above)
Comment (valeriocos) [reply perceval/client.py]: get methods should pass the payload by the argument **params**, while post methods by the argument **data**. Btw, I guess the behavior is the same, maybe a couple of tests will need to be changed (If I remember well, Bugzilla or Bugzillarest).

File: perceval/client.py (hunk 3 above)
Comment (sduenas) [reply perceval/client.py]: But, why do we need to do it this way? Why don't you need to modify Retry object and not the instance of this object?

File: perceval/client.py (hunk 1 above)
Comment (sduenas) [reply perceval/client.py]: I see, but I think the docs make reference to parameters needed by the server that won't be harmful, like in Bugzilla with the session cookies. The server may define other cookies but in my opinion, it won't be a problem. And it will keep the configuration for retrying without having to define it each time a requests is sent.

File: perceval/client.py (hunk 2 above)
Comment (sduenas) [reply perceval/client.py]: Yep, you're right. Btw, what's `stream` parameter and why do we need it?

File: perceval/client.py (hunk 4 above)
Comment (sduenas) [reply perceval/client.py]: That's the idea. To create a session, to define which parameters are needed and fetch the items while connection or defined HTTP errors are handled to retry.

When at first I thought of handling rate limit in a generic client, it looked a good idea to me. But now I see it on the code, I feel it a bit weird. So, having a specific class which handles only those cases it's a better idea to me.

What do you think?

File: perceval/client.py (hunk 2 above)
Comment (valeriocos) [reply perceval/client.py]: The stream parameter is used in **pipermail** and **hyperkitty** clients. It defers downloading the response body until you access the **Response.content** attribute (the default behavior is to immediately download the body of the response).

File: perceval/client.py (hunk 4 above)
Comment (valeriocos) [reply perceval/client.py]: Let's try with a specific class

File: perceval/client.py (hunk 3 above)
Comment (valeriocos) [reply perceval/client.py]: You're right, better to modify the instance of the object

File: perceval/client.py (hunk 2 above)
Comment (sduenas) [reply perceval/client.py]: Great. I forgot about it!

File: perceval/client.py (hunk 3 above)
Comment (valeriocos) [reply perceval/client.py]: I tried to modify the instance of the object, but the new error status codes were not taken into account (the test **test_abuse_rate_limit** was failing). So I left the code as it was

File: perceval/client.py (hunk 4 above)
Comment (valeriocos) [reply perceval/client.py]: I added the class **RateLimitHandler** in the client.py. 
The github client remains a subclass of HttpClient -> **GitHubClient(HttpClient)**, while the HttpClient can accept a rate-limit-handler as parameter, which allows to deal with rate limit issues.

This is a bit different from your proposal (**GitHubClient(RateLimitHandler, HttpClient)**), but I'd would say that the generic handler should be coupled with the generic client. In case a given client wants to use the handler, it will have to set up the handler and pass it to the generic client, like this:

```python
rate_limit_handler = RateLimitHandler(sleep_for_rate, min_rate_to_sleep)
super().__init__(self.api_url, default_sleep_time=default_sleep_time, max_retries=max_retries,
                 rate_limit_handler=rate_limit_handler)
```

your thoughts?

File: perceval/client.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        yield self._fetch(url, payload, headers, method, stream,
+                          retry_on_status_codes,
+                          retry_after_status_code)
+
+    def _fetch(self, url, payload=None, headers=None, method=GET,
+               stream=False,
+               retry_on_status_codes=None,
+               retry_after_status_codes=None):
+
+        self.sleep_for_rate_limit()
```
Comment (sduenas) [inline perceval/client.py]: Instead og having this code here, I would do the next, taking advantage of multiple inheritance:
Let's take `GitHub` class the base case:

```
class GitHub(HttpClient, RateLimitHandler):

    def fetch(self, url, payload=None, headers=None, method=GET, retry_on_status_codes=None, retry_after_status_code=None):
        self.sleep_for_rate_limit()
        yield super().fetch(url, payload, headers, method, stream, retry_on_status_codes, retry_after_status_code)
        self.update_rate_limit(response)
```

Thus,  `sleep_for_rate_limit` and `update_rate_limit` can be safely removed from `HttpClient`. In addition, GitHub client `__init__` will need to call the `__init__` method of both parets.

https://stackoverflow.com/questions/26927571/multiple-inheritance-in-python3-with-different-signatures

Maybe we can do the same with pagination? What do you think?

File: perceval/client.py [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+        yield self._fetch(url, payload, headers, method, stream,
+                          retry_on_status_codes,
+                          retry_after_status_code)
+
+    def _fetch(self, url, payload=None, headers=None, method=GET,
+               stream=False,
+               retry_on_status_codes=None,
+               retry_after_status_codes=None):
+
+        self.set_retry_after_status_codes(retry_after_status_codes)
```
Comment (sduenas) [inline perceval/client.py]: I would move all this retrying setup to `create_http_session()`.

File: perceval/client.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+        if headers:
+            session.headers.update(headers)
+
+        return session
+
+    def close_http_session(self):
+        self.session.close()
+
+    def set_retry_after_status_codes(self, status_codes=None):
```
Comment (sduenas) [inline perceval/client.py]: I still don't see why this is needed. I mean, why can't these parameters being set during Retry initialization in line 107?

File: perceval/client.py (hunk 6 above)
Comment (valeriocos) [reply perceval/client.py]: if we move it to _create_http_session()_, we won't be able to control every call to __fetch_. Imagine that for a given call we want to set the retry after for a 403, but in another call we don't want.

Btw, I admit I haven't found any example in the backends where this happens

File: perceval/client.py (hunk 7 above)
Comment (valeriocos) [reply perceval/client.py]: I don't see any parameter to set the retry-after-status-codes during the initialization of the object. 
I have tried to set them up just after the initialization (line 118) but I didn't succeed.

I'm looking at [urllib3](http://urllib3.readthedocs.io/en/latest/reference/urllib3.util.html)

File: perceval/client.py [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+#
+# Authors:
+#     Valerio Cosentino <valcos@bitergia.com>
+#
+
+import logging
+import time
+
+import requests
+from requests.adapters import HTTPAdapter
```
Comment (sduenas) [inline perceval/client.py]: Use only `import requests` and `import urllib3.util`. And use them when calling to `HTTPAdapter` or `Retry`. It's clearer.

File: perceval/client.py [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+
+from .errors import RateLimitError
+
+logger = logging.getLogger(__name__)
+
+
+class HttpClient:
+    """Abstract class for HTTP clients.
+
+    Base class to query data sources taking care of retrying requests in case connection issues. If the data source does
```
Comment (sduenas) [inline perceval/client.py]: Make these lines shorter and update the documentation. There are references to methods like `_fetch` which are no longer in use.

File: perceval/client.py [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+    DEFAULT_RAISE_ON_STATUS = True
+    DEFAULT_RESPECT_RETRY_AFTER_HEADER = True
+
+    DEFAULT_RETRY_AFTER_STATUS_CODES = [413, 429, 503]
+    DEFAULT_STATUS_FORCE_LIST = [408, 423, 504]
+
+    GET = "GET"
+    POST = "POST"
+
+    def __init__(self, base_url, max_retries=MAX_RETRIES, max_retries_on_connect=MAX_RETRIES_ON_CONNECT,
```
Comment (sduenas) [inline perceval/client.py]: Do we really need all these configuration parameters? I think we should stick to only those we will use for sure. If we need more, we will add them in future versions.

File: perceval/client.py [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+        self.raise_on_status = raise_on_status
+        self.respect_retry_after_header = respect_retry_after_header
+        self.default_sleep_time = default_sleep_time
+
+        self._create_http_session(headers)
+
+    def __del__(self):
+        self._close_http_session()
+
+    def _create_http_session(self, headers=None):
```
Comment (sduenas) [inline perceval/client.py]: Private methods go after public methods.

File: perceval/client.py [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+        """
+        Create a http session and initialize the retry object
```
Comment (sduenas) [inline perceval/client.py]: Please update the docstrings to the style we have in other parts of the code.

File: perceval/client.py [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+
+        self.session.mount('http://', HTTPAdapter(max_retries=retries))
+        self.session.mount('https://', HTTPAdapter(max_retries=retries))
+
+    def _close_http_session(self):
+        """Close the http session"""
+
+        self.session.close()
+
+    def fetch(self, url, payload=None, headers=None, method=GET, stream=False):
```
Comment (sduenas) [inline perceval/client.py]: Same comment as above with the documentation.

File: perceval/client.py [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+        """
+        Fetch the data from a given URL
+
+        :param url: link to the resource
+        :param payload: payload of the request
+        :param headers: headers of the request
+        :param method: type of request call (GET or POST)
+        :param stream: defer downloading the response body until the response content is available
+        """
+        return self.__send_request(url, payload, headers, method, stream)
```
Comment (sduenas) [inline perceval/client.py]: As this method doesn't do anything else than calling `_send_request`, they should be merged into one.

File: perceval/client.py [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+        else:
+            response = self.session.post(url, data=payload, headers=headers, stream=stream)
+
+        response.raise_for_status()
+
+        return response
+
+
+class RateLimitHandler:
+    """Class to handle rate limit for HTTP clients.
```
Comment (sduenas) [inline perceval/client.py]: Check documentation style all along this class.

File: tests/test_client.py [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+    def __init__(self, base_url, default_sleep_time=0.1, max_retries=1):
+
+        super().__init__(base_url, default_sleep_time=default_sleep_time, max_retries=max_retries)
+        super().setup_rate_limit_handler()
+
+
+class TestHttpClient(unittest.TestCase):
+    """Http client tests """
+
+    def test_initialization(self):
```
Comment (sduenas) [inline tests/test_client.py]: We need another test where we check that the parameters passed to the class are well set. I mean, there aren't assignment errors to variables.

File: tests/test_client.py [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+                               CLIENT_SPIDERMAN_URL,
+                               body=output,
+                               status=200)
+
+        client = MockedClient(CLIENT_API_URL)
+        response = client.fetch(CLIENT_SPIDERMAN_URL)
+        self.assertEqual(response.headers['connection'], 'close')
+
+    @httpretty.activate
+    def test_fetch(self):
```
Comment (sduenas) [inline tests/test_client.py]: Slipt these checks into three independent tests. Therefore, a failure in one of the tests doesn't stop for testing the others.

File: tests/test_client.py [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+        client = MockedClient(CLIENT_API_URL)
+
+        urls = [CLIENT_SPIDERMAN_URL, CLIENT_SUPERMAN_URL, CLIENT_BATMAN_URL]
+
+        for url in urls:
+            with self.assertRaises(requests.exceptions.RetryError):
+                _ = client.fetch(url)
+
+    @httpretty.activate
+    def test_init_rate_limit(self):
```
Comment (sduenas) [inline tests/test_client.py]: Rate limit these tests should go in a different test class. I also miss some tests here.

**Comment by sduenas** [comment]
A newer version of this PR has been merged. Closing it.

**Comment by zhuqianbin** [comment]
Hi, How to let perceval support cookie? Since some web site need to establish a connection with cookie.

**Comment by jgbarah** [comment]
@zhuqianbin y saw your comment by chance. Please, open an issue if you have a question like this (not a comment in a pull request). If you do, please clarify what do you mean. Which perceval backend (or data source) you intend to use, and which kind of cookie you are considering (because as far as I know, none of the data sources supported by Perceval use cookies).
